internal/expr/expr_test.go

```diff
@@ -293,281 +233,195 @@ var tests = []struct {
 	}, {
-		sql:      "SELECT street AS &Address.street FROM t",
-		structs:  []any{},
-		errorstr: "cannot prepare expression: type Address unknown, have: ",
+		query:       "SELECT p.name AS (&Address.district, &Address.street) FROM t",
+		prepareArgs: []any{Address{}},
+		err:         "cannot prepare expression: mismatched number of cols and targets in output expression: p.name AS (&Address.district, &Address.street)",
 	}, {
-		sql:      "SELECT street AS &Address.id FROM t",
-		structs:  []any{Person{ID: 1}},
-		errorstr: "cannot prepare expression: type Address unknown, have: Person",
-	}}
-
-	for i, test := range testList {
-		parser := expr.NewParser()
-		parsedExpr, err := parser.Parse(test.sql)
-		if err != nil {
-			c.Fatal(err)
-		}
-		_, err = parsedExpr.Prepare(test.structs...)
-		c.Assert(err, ErrorMatches, test.errorstr,
-			Commentf("test %d failed:\nsql: '%s'\nstructs: '%+v'", i, test.sql, test.structs))
-	}
-
-}
-
-func (s *ExprSuite) TestPrepareMissingTag(c *C) {
-	testList := []struct {
-		sql      string
-		structs  []any
-		errorstr string
-	}{{
-		sql:      "SELECT street FROM t WHERE x = $Address.number",
-		structs:  []any{Address{ID: 1}},
-		errorstr: `cannot prepare expression: type Address has no "number" db tag`,
+		query:       "SELECT (&Address.*, &Address.id) FROM t",
+		prepareArgs: []any{Address{}, Person{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (&Address.*, &Address.id)",
 	}, {
-		sql:      "SELECT (street, road) AS &Address.* FROM t",
-		structs:  []any{Address{ID: 1}},
-		errorstr: `cannot prepare expression: type Address has no "road" db tag`,
+		query:       "SELECT (p.*, t.name) AS &Address.* FROM t",
+		prepareArgs: []any{Address{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (p.*, t.name) AS &Address.*",
 	}, {
-		sql:      "SELECT &Address.road FROM t",
-		structs:  []any{Address{ID: 1}},
-		errorstr: `cannot prepare expression: type Address has no "road" db tag`,
-	}}
-
-	for i, test := range testList {
-		parser := expr.NewParser()
-		parsedExpr, err := parser.Parse(test.sql)
-		if err != nil {
-			c.Fatal(err)
-		}
-		_, err = parsedExpr.Prepare(test.structs...)
-		c.Assert(err, ErrorMatches, test.errorstr,
-			Commentf("test %d failed:\nsql: '%s'\nstructs:'%+v'", i, test.sql, test.structs))
-
-	}
-}
-
-func (s *ExprSuite) TestPrepareInvalidAsteriskPlacement(c *C) {
-	testList := []struct {
-		sql      string
-		structs  []any
-		errorstr string
-	}{{
-		sql:      "SELECT (&Person.*, &Person.*) FROM t",
-		structs:  []any{Address{}, Person{}},
-		errorstr: "cannot prepare expression: invalid asterisk in output expression: Output[[] [Person.* Person.*]]",
+		query:       "SELECT (name, p.*) AS (&Person.id, &Person.*) FROM t",
+		prepareArgs: []any{Address{}, Person{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (name, p.*) AS (&Person.id, &Person.*)",
```

Comment:
These messages are a lot nicer, thank you.  I wonder if they'll get a bit out of hand when the expressions get larger, but that's a problem we already had and we can always turn the expression into a suffix+"..." after a given length if it becomes awkward.

Reply:
True, though hopefully most the time the separate output structs will be in seperate output expressions and we shouldn't end up with massive chunks of text here.